DOCS-4086: code samples for building a good dataset (#4413)
viamrobotics/docs: docs/data-ai/train/update-dataset.md

Author: Moved the overlapping content onto this page. Ideally that example belongs on a separate (advanced?) how-to page that covers programmatic capture, tagging, and dataset creation. But I think for now it works in the 'add to' scope, since adding comes last in the narrative.

---
linkTitle: "Add to a training dataset"

> title: "Add to a training dataset"

Reviewer: ```suggestion
title: "Add images to a training dataset"
```

weight: 30
layout: "docs"
type: "docs"

> description: "Update a dataset used to train a machine learning model."

Reviewer: ```suggestion
description: "Update a dataset with more images to train a machine learning model."
```

```typescript
const client: ViamClient = await createViamClient({
  credential: {
    type: "api-key",
    payload: API_KEY,
  },
  authEntity: API_KEY_ID,
});

const dataClient = client.dataClient;

// Add image to dataset
await dataClient.addBinaryDataToDatasetByIds({
  binaryIds: [EXISTING_IMAGE_ID],
  datasetId: EXISTING_DATASET_ID,
});

console.log("Image added to dataset successfully");
client.disconnect();
```

{{% /tab %}}
{{% tab name="Flutter" %}}

To add an image to a dataset, find the binary data ID for the image and the dataset ID.
Pass both IDs to [`dataClient.addBinaryDataToDatasetByIDs`](/dev/reference/apis/data-client/#addbinarydatatodatasetbyids):

```dart
final client = await ViamClient.withApiKey(
    apiKeyId: apiKeyId,
    apiKey: apiKey,
);

final dataClient = client.dataClient;

try {
    // Add image to dataset
    await dataClient.addBinaryDataToDatasetByIds(
      binaryIds: [existingImageId],
      datasetId: existingDatasetId,
    );

    print('Image added to dataset successfully');
} finally {
    await client.close();
}
```

{{% /tab %}}
{{< /tabs >}}

> ## Add all images captured by a specific machine to a dataset

Author: I wanted to avoid documenting the web UI since it's more subject to change, but that's a great point. I'll add a tab for the web UI here, avoiding screenshots ofc

The following script adds all images captured from a certain machine to a new dataset:

{{< tabs >}}

> {{% tab name="Python SDK" %}}

Reviewer: ```suggestion
{{% tab name="Python" %}}
```
[nit] consistency

        if not data:
            break
        all_matches.extend(data)

    return all_matches

async def main() -> int:
    """Main execution function."""
    viam_client = await connect()
    data_client = viam_client.data_client

    matching_data = await fetch_binary_data_ids(data_client, PART_ID)

    print("Creating dataset...")

    try:
        dataset_id = await data_client.create_dataset(
            name=DATASET_NAME,
            organization_id=ORG_ID,
        )
        print(f"Created dataset: {dataset_id}")
    except Exception as e:
        print("Error creating dataset. It may already exist.")
        print("See: https://app.viam.com/data/datasets")
        print(f"Exception: {e}")
        return 1

    print("Adding data to dataset...")

    await data_client.add_binary_data_to_dataset_by_ids(
        binary_ids=[
            obj.metadata.binary_data_id for obj in matching_data
        ],
        dataset_id=dataset_id
    )

    print("Added files to dataset.")
    print(
        f"See dataset: https://app.viam.com/data/datasets?id={dataset_id}"
    )

    viam_client.close()
    return 0

if __name__ == "__main__":
    asyncio.run(main())
```

{{% /tab %}}

> {{% tab name="Go SDK" %}}

Reviewer: ```suggestion
{{% tab name="Go" %}}
```
[nit] consistency

    if err != nil {
        log.Fatalf("Failed to fetch binary data: %v", err)
    }

    fmt.Println("Creating dataset...")

    datasetResp, err := dataClient.CreateDataset(ctx, &data.CreateDatasetRequest{
        Name:           DATASET_NAME,
        OrganizationId: ORG_ID,
    })
    if err != nil {
        fmt.Println("Error creating dataset. It may already exist.")
        fmt.Println("See: https://app.viam.com/data/datasets")
        fmt.Printf("Exception: %v\n", err)
        return
    }

    datasetID := datasetResp.Id
    fmt.Printf("Created dataset: %s\n", datasetID)

    fmt.Println("Adding data to dataset...")

    var binaryIDs []string
    for _, obj := range matchingData {
        binaryIDs = append(binaryIDs, obj.Metadata.Id)
    }

    _, err = dataClient.AddBinaryDataToDatasetByIds(ctx, &data.AddBinaryDataToDatasetByIdsRequest{
        BinaryIds: binaryIDs,
        DatasetId: datasetID,
    })
    if err != nil {
        log.Fatalf("Failed to add binary data to dataset: %v", err)
    }

    fmt.Println("Added files to dataset.")
    fmt.Printf("See dataset: https://app.viam.com/data/datasets?id=%s\n", datasetID)
}

```

{{% /tab %}}

> {{% tab name="Typescript" %}}

Reviewer: ```suggestion
{{% tab name="TypeScript" %}}
```
[nit]

    );

    if (response.data.isEmpty) {
      break;
    }

    allMatches.addAll(response.data);
    last = response.last;
  }

  return allMatches;
}

Future<int> main() async {
  final viamClient = await connect();
  final dataClient = viamClient.dataClient;

  final matchingData = await fetchBinaryDataIds(dataClient, partId);

  print("Creating dataset...");

  try {
    final datasetId = await dataClient.createDataset(
      name: datasetName,
      organizationId: orgId,
    );
    print("Created dataset: $datasetId");

    print("Adding data to dataset...");

    final binaryIds = matchingData
        .map((obj) => obj.metadata.binaryDataId)
        .toList();

    await dataClient.addBinaryDataToDatasetByIds(
      binaryIds: binaryIds,
      datasetId: datasetId,
    );

    print("Added files to dataset.");
    print("See dataset: https://app.viam.com/data/datasets?id=$datasetId");

    viamClient.close();
    return 0;

  } catch (error) {
    print("Error creating dataset. It may already exist.");
    print("See: https://app.viam.com/data/datasets");
    print("Exception: $error");
    viamClient.close();
    return 1;
  }
}
```

{{% /tab %}}
{{< /tabs >}}

> ## Capture, annotate, and add images to a dataset

Author: Reorganized according to your other comment; hopefully that helps!

> ## Capture, annotate, and add images to a dataset

Author: noting to self that we can significantly simplify this code when uploadfiletodataset happens

        return nil, fmt.Errorf("API credentials not configured")
    }

    client, err := appclient.New(ctx, appclient.Config{
        KeyID: dc.apiKeyID,
        Key:   dc.apiKey,
    })
    if err != nil {
        return nil, fmt.Errorf("failed to create app client: %w", err)
    }

    return client.DataClient, nil
}

func (dc *DataCollector) encodeImageToPNG(img image.Image) ([]byte, error) {
    // PNG encoding implementation
    return nil, nil // Placeholder
}

func NewDataCollector(logger logging.Logger, componentName, datasetID, apiKeyID, apiKey string) (*DataCollector, error) {
    if logger == nil {
        return nil, fmt.Errorf("logger is required")
    }
    if componentName == "" {
        return nil, fmt.Errorf("component name is required")
    }
    if datasetID == "" {
        return nil, fmt.Errorf("dataset ID is required")
    }
    if apiKeyID == "" || apiKey == "" {
        return nil, fmt.Errorf("API credentials are required")
    }

    return &DataCollector{
        logger:        logger,
        componentName: componentName,
        datasetID:     datasetID,
        apiKeyID:      apiKeyID,
        apiKey:        apiKey,
    }, nil
}
```

{{% /tab %}}

> {{% tab name="Typescript" %}}

Reviewer: ```suggestion
{{% tab name="TypeScript" %}}
```
[nit]
